classes/audit_manager.php

```diff
@@ -0,0 +1,376 @@
+     *
+     * @param int $limitfrom return a subset of records, starting at this point (optional).
+     * @param int $limitnum return a subset comprising this many records (optional, required if $limitfrom is set).
+     *
+     * @return \moodle_recordset A moodle_recordset instance
+     * @throws \dml_exception
+     * @throws \moodle_exception
+     */
+    public static function get_audit_results_paginated($limitfrom=0, $limitnum=0) {
+        global $DB;
+
+        $recordset = $DB->get_recordset('report_apocalypse', null, '', '*', $limitfrom, $limitnum);
+
+        $activities = self::get_instances($recordset);
+
+        return $activities;
+    }
+
+    /**
+     * Helper method to convert db records to audit_activity objects.
+     *
+     * @param \moodle_recordset $records iterative containing mixed fieldset objects.
+     *
+     * @return array of audit_activity objects.
+     * @throws \moodle_exception
+     */
+    public static function get_instances(\moodle_recordset $records) {
+        $activities = array();
+        foreach ($records as $activity) {
+            $activities[] = new audit_activity($activity);
+        }
+        return $activities;
+    }
+
+    /**
+     * Delete previous audit results from database.
+     *
+     * @throws \dml_exception
+     */
+    public static function delete_records() {
+        global $DB;
+
+        $DB->delete_records('report_apocalypse');
+
+    }
+
+    /**
+     * Store passed in recordset and a transaction record in database.
+     *
+     * @param \moodle_recordset $recordset Records to store
+     *
+     * @return bool true if successful, false if not.
+     * @throws \Exception
+     */
+    public static function store_records($recordset) {
+
+        if ($recordset->valid()) {
+            $count = 0;
+
+            foreach ($recordset as $record) {
+                self::insert_audit_activity($record);
+                $count++;
+            }
+            $recordset->close();
+            self::insert_audit_record($count);
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    /**
+     * Get the details of the last time a flash audit was run.
+     */
+    public static function get_last_audit() {
+        global $DB;
+
+        $sql = 'SELECT * FROM {report_apocalypse_audits}
+                WHERE rundatetime = (SELECT MAX(rundatetime) FROM {report_apocalypse_audits})';
+        return $DB->get_record_sql($sql);
+    }
+
+    /**
+     * Iterate over a \moodle_recordset and insert values into 'report_apocalypse' table.
+     *
+     * @param mixed $record a fieldset object containing a record
+     *
+     * @throws \dml_exception
+     * @throws \moodle_exception
+     */
+    public static function insert_audit_activity($record) {
+        global $DB;
+
+        $activity = self::build_audit_activity_from_record($record);
+        $DB->insert_record('report_apocalypse', $activity->get_activity(), false);
```

Comment:
you should be restructuring this to make use of the insert_records api calls instead of inserting a single record at a time.

Reply:
I removed this function and utilised the `insert_records` api call.